library/general/src/lib/yast2/log_dir_rotator.rb

```diff
@@ -0,0 +1,195 @@
+    # and delete. Child classes should override the value.
+    NAME = "undefined".freeze
+
+    # Prepare (create, rotate and delete) the log directory.
+    def prepare
+      return if Yast::Mode.test
+      begin
+        log.info "preparing log dir #{self.class::NAME}"
+        rotate_log_dirs
+        FileUtils.mkdir_p(log_dir)
+      rescue SystemCallError => e
+        log.error e.to_s
+      end
+    end
+
+    # Copy a file to the log directory.
+    #
+    # @param src [String] name of source
+    # @param dest [String] basename of destination
+    def copy(src, dest)
+      return if Yast::Mode.test
+      begin
+        FileUtils.cp(src, log_dir + "/" + dest)
+      rescue SystemCallError => e
+        log.error e.to_s
+      end
+    end
+
+    # Write a file in the log directory.
+    #
+    # @param dest [String] basename of destination
+    # @param content [String] content to write
+    def write(dest, content)
+      return if Yast::Mode.test
+      begin
+        File.open(log_dir + "/" + dest, "w") { |file| file.write(content) }
+      rescue SystemCallError => e
+        log.error e.to_s
+      end
+    end
+
+    # Return a suitable name for the devicegraph log directory
+    # depending on the YaST mode (installation / installed system).
+    #
+    # @return [String] directory name with full path
+    def log_dir
+      dir = installation? ? "#{self.class::NAME}-inst" : self.class::NAME
+      base_dir + "/" + dir
+    end
+
+  private
+
+    # Return true if this is some installation mode: installation, update
+    #
+    # @return [Boolean]
+    def installation?
+      Yast::Mode.installation || Yast::Mode.update
+    end
+
+    # Rotate the log directories, depending on current YaST mode:
+    #
+    # During installation (or update or AutoYaST), clear and remove any old
+    # /var/log/YaST2/#{NAME}-inst directory.
+    #
+    # In the installed system, keep a number of old log directories, remove
+    # any older ones in /var/log/YaST2, and rename the ones to keep:
+    #
+    #   rm -rf #{NAME}-03
+    #   mv #{NAME}-02 #{NAME}-03
+    #   mv #{NAME}-01 #{NAME}-02
+    #   mv #{NAME}    #{NAME}-01
+    #
+    # This will NOT create any new log directory.
+    def rotate_log_dirs
+      return unless File.exist?(base_dir)
+      if installation?
+        kill_old_log_dirs([File.basename(log_dir)])
+      else
+        log_dirs = old_log_dirs.sort
+        keep_dirs = log_dirs.shift(self.class::KEEP_OLD_LOG_DIRS)
+        kill_old_log_dirs(log_dirs)
+        keep_dirs.reverse.each { |dir| rename_old_log_dir(dir) }
+      end
+    end
+
+    # Kill (recursively remove) old log directories.
+    #
+    # @param log_dirs [Array<String>] directory names (without path) to remove
+    def kill_old_log_dirs(log_dirs)
+      log_dirs.each do |dir|
+        next unless File.exist?(base_dir + "/" + dir)
+        log.info("Removing old log dir #{dir}")
+        FileUtils.remove_dir(base_dir + "/" + dir)
```

Comment:
this will potentially raise exception.  I suggest to use `FileUtils.remove_dir(base_dir + "/" + dir, true)`. See https://ruby-doc.org/stdlib-2.4.1/libdoc/fileutils/rdoc/FileUtils.html#method-c-remove_dir

Reply:
Exceptions are logged as documented.